### ISSUE TITLE
Support headers on `ask` endpoint requests

### ISSUE THREAD

**Issue by spwitt** [caddyserver/caddy#5562]
Add support for including headers in the `ask` endpoint request.

Expand the `on_demand_tls` directive to allow specifying one or more headers in the form of key-value-pairs. These key-value-pairs will be added to the `ask` endpoint request.

One use case for this feature is supporting authentication on the `ask` endpoint, but the more generic key-value-pair approach allows the most flexibility for feature use.

**Comment by mholt** [comment]
Thanks for the feature request.

Can you expand on your need for this? Why do you need to authenticate the 'ask' endpoint? It should usually be internal-only anyways.

Why is anything else other than the domain name needed for a yes/no answer?

**Comment by francislavoie** [comment]
The ask endpoint should always be super-fast (i.e. `O(1)` lookup in an indexed mapping/DB) and the result is not sensitive information because domains are public information (e.g. tracked in certificate transparency logs). There's no harm leaving that endpoint unauthenticated.

It's kinda like a registration form with an "email already in use" validation, you don't authenticate validation on a registration form. Same idea here, it's just a validation request.

**Comment by emilylange** [comment]
You could also, theoretically, have the `ask` endpoint be just another vhost in your Caddyfile (and independent of other vhosts, e.g. via [`bind`](https://caddyserver.com/docs/caddyfile/directives/bind)).
And within that vhost, you could do all kinds of funky mangling, inject or validate auth, and whatnot and eventually just [`reverse_proxy`](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy) to your actual `ask` endpoint.

Just as a fyi, thinking a bit out of the box.
Though, be aware that it comes with its own downsides. 

But I do very much agree with @francislavoie :)

**Comment by spwitt** [comment]
Thank you all for the prompt response!

We had the same discussions internally about whether the `ask` endpoint needed to require an API key for access. We ended up using something like @emilylange and @francislavoie suggested with the reverse proxy even before I saw this response, so at least we're thinking on the right track.

We'll use that reverse proxy for now unless it seems the API authentication is causing issues.
